src/style/style_layer/color_relief_style_layer.ts

```
@@ -0,0 +1,45 @@
+import {StyleLayer} from '../style_layer';
+
+import properties, {type ColorReliefPaintPropsPossiblyEvaluated} from './color_relief_style_layer_properties.g';
+import {type Transitionable, type Transitioning, type PossiblyEvaluated} from '../properties';
+
+import type {ColorReliefPaintProps} from './color_relief_style_layer_properties.g';
+import {type Color, Interpolate, ZoomConstantExpression, type LayerSpecification} from '@maplibre/maplibre-gl-style-spec';
+import {nextPowerOfTwo} from '../../util/util';
+
+export const isColorReliefStyleLayer = (layer: StyleLayer): layer is ColorReliefStyleLayer => layer.type === 'color-relief';
+
+export class ColorReliefStyleLayer extends StyleLayer {
+    elevationStops: Array<number>;
+    colorStops: Array<Color>;
+    _transitionablePaint: Transitionable<ColorReliefPaintProps>;
+    _transitioningPaint: Transitioning<ColorReliefPaintProps>;
+    paint: PossiblyEvaluated<ColorReliefPaintProps, ColorReliefPaintPropsPossiblyEvaluated>;
+
+    constructor(layer: LayerSpecification) {
+        super(layer, properties);
+        this._updateColorRamp();
+    }
+
+    _updateColorRamp() {
+        const expression = this._transitionablePaint._values['color-relief-color'].value.expression;
+        if (expression instanceof ZoomConstantExpression && expression._styleExpression.expression instanceof Interpolate) {
```

What to do if the expression is not an `interpolate` expression` that uses `elevation`? Currently, we silently fail to display the layer. `HeatmapStyleLayer` has similar requirements, and does not seem to validate its input. Should we throw an error here in the constructor? Or display a default? Or something else?

---

It is possible not to check the expression type? I prefer to have this less specific and more future-proof. I think we should imitate how it is in the heatmap layer.

---

The current implementation uses the color ramp in GLSL. So rather than using the public `Expression` interface (like Heatmap does in `renderColorRamp()`), I'm extracting the elevation and color tables from the `Expression` internals and passing them to the shader.

`Heatmap` evaluates the color ramp onto a uniform grid (using `expression.evaluate()` and then passes the rendered color ramp to the shader as a texture. The shader can then do a simple, fast texture sample to get the appropriate color. This is fast and works well for smoothly-varying color ramps, and was my initial implementation for `color-relief`. However, it does not support sharp breaks in the color ramp, which is a LINZ requirement (see "LINZ" and "categorical" style examples in the [demo](https://nathanmolson.github.io/color_relief)).

The specific line that requires the expression to be an `Interpolate` is line 28,

```
this.elevationStops = interpolater.labels;
```

This is used to extract the elevation table out of the expression, so it can be used in the shader.